---
title: "Create data package"
author: "Trang Tran"
date: "March 22, 2020"
output: html_notebook
---

```{r setup, include=FALSE, echo=FALSE}
library(magrittr)
library(ggplot2)
library(Biobase)

options(stringsAsFactors = FALSE)
knitr::opts_chunk$set(echo = TRUE)
```

## Overview

## Download count matrices and meta data, set environment variable 'DBDIR'

```{r,eval = FALSE}
Sys.setenv('DBDIR' = '.')
```

It is assumed that files are organized into directories corresponding to the workflow, i.e. `kallisto`, `star-rsem`, `subread`.

## Experiment description - same for all data sets

```{r}
experimentData <- new("MIAME",
                      lab = "Center for Pharmacogenomics",
                      contact = "leming.shi@gmail.com",
                      title = "SEQC Toxicogenomics Study: RNA-Seq data set",
                      abstract = "",
                      other = list(
                          citation = "Gong B, Wang C, Su Z, Hong H et al. Transcriptomic profiling of rat liver samples in a comprehensive study design by RNA-Seq. Sci Data 2014;1:140021. PMID: 25977778"
                      )
                      )
```


## Kallisto workflow

### Gene-level `ExpressionSet`

#### Assay data

```{r}
gene.counts = readRDS(file.path(Sys.getenv('DBDIR'), 'kallisto', 'matrix.gene.est_counts.RDS'))
gene.tpm = readRDS(file.path(Sys.getenv('DBDIR'), 'kallisto', 'matrix.gene.tpm.RDS'))
gene.assayData <- new.env()
assign('exprs', gene.counts, envir = gene.assayData) # exprs is required
assign('count', gene.counts, envir = gene.assayData)
assign('tpm', gene.tpm, envir = gene.assayData)
ExpressionSet(assayData = gene.assayData)
```

#### Phenotypic data

```{r}
metadata = read.table(file.path(Sys.getenv('DBDIR'),
                                'PRJNA239561_metadata_cleaned.tsv'), sep = '\t', header = TRUE)
libnames <- colnames(gene.assayData$exprs) %>% as.character()
phenoData <- data.frame('RunId' = libnames) %>%
    plyr::join(y = metadata, by = 'RunId', type = 'left')  %>%
    set_rownames(libnames) %>%
    Biobase::AnnotatedDataFrame(data = ., varMetadata = data.frame('labelDescription' = colnames(.), row.names = colnames(.)))   
```

#### Annotations and features

```{r}
feature_attrs = read.table(file.path(Sys.getenv('DBDIR'), 'kallisto',
                                     'feature_attributes.tsv'),
                           sep = '\t', header=TRUE, colClasses = 'character')
is.gene.NA = which(is.na(feature_attrs$gene_id))
feature_attrs[is.gene.NA, 'gene_id'] = feature_attrs[is.gene.NA, 'transcript_id']
gene.featureData <- data.frame('gene_id' = rownames(gene.assayData$exprs)) %>%
    plyr::join(y = feature_attrs[, c('gene_id', 'gene_biotype', 'gene_symbol')], by = 'gene_id', type = 'left') %>%
    set_names(c('ID', 'Biotype', 'Symbol')) %>%
    unique() 
rownames(gene.featureData) = gene.featureData$ID
gene.featureData = gene.featureData %>%
    Biobase::AnnotatedDataFrame(data = ., varMetadata = data.frame('labelDescription' = colnames(.), row.names = colnames(.)))
```

#### Experiment description

Generated earlier, same for all data sets.

#### Assembling an `ExpressionSet`

```{r}
rnor.rnaseq.gene.kallisto <- ExpressionSet(assayData = gene.assayData,
                             phenoData = phenoData,
                             experimentData = experimentData,
                             featureData = gene.featureData,
                             annotation = 'Ensembl Rnor_6.0.99 cdna + ERCC-92'
                             )
```

### Transcript-level `ExpressionSet`

#### Assay data

```{r}
transcript.counts = readRDS(file.path(Sys.getenv('DBDIR'), 'kallisto', 'matrix.est_counts.RDS'))
transcript.tpm = readRDS(file.path(Sys.getenv('DBDIR'), 'kallisto', 'matrix.tpm.RDS'))
transcript.assayData <- new.env()
assign('exprs', transcript.counts, envir = transcript.assayData) # exprs is required
assign('count', transcript.counts, envir = transcript.assayData)
assign('tpm', transcript.tpm, envir = transcript.assayData)
ExpressionSet(assayData = transcript.assayData)
```

#### Phenotypic data

Same as above

#### Annotations and features

```{r}
# feature_attrs = read.table(file.path(Sys.getenv('DBDIR'), 'feature_attributes.tsv'), sep = '\t', header=TRUE, colClasses = 'character')
transcript.featureData <- data.frame('transcript_id' = rownames(transcript.assayData$exprs)) %>%
    plyr::join(y = feature_attrs[, c('transcript_id', 'location', 'transcript_biotype')], by = 'transcript_id', type = 'left') %>%
    set_names(c('ID', 'Location', 'Biotype'))
rownames(transcript.featureData) = transcript.featureData$ID
transcript.featureData = transcript.featureData %>%
    Biobase::AnnotatedDataFrame(data = ., varMetadata = data.frame('labelDescription' = colnames(.), row.names = colnames(.)))
```

#### Experiment description

Same as above

#### Assembling an `ExpressionSet`

```{r}
rnor.rnaseq.transcript.kallisto <- ExpressionSet(assayData = transcript.assayData,
                             phenoData = phenoData,
                             experimentData = experimentData,
                             featureData = transcript.featureData,
                             annotation = 'Ensembl Rnor_6.0.99 cdna + ERCC-92'
                             )
```



## STAR-RSEM workflow

### Gene-level `ExpressionSet`

#### Assay data

```{r}
gene.counts = readRDS(file.path(Sys.getenv('DBDIR'), 'star-rsem', 'matrix.gene.expected_count.RDS'))
gene.tpm = readRDS(file.path(Sys.getenv('DBDIR'), 'star-rsem', 'matrix.gene.tpm.RDS'))
gene.assayData <- new.env()
assign('exprs', gene.counts, envir = gene.assayData) # exprs is required
assign('expected_count', gene.counts, envir = gene.assayData)
assign('tpm', gene.tpm, envir = gene.assayData)
ExpressionSet(assayData = gene.assayData)
```

#### Annotations and features

```{r}
gene.featureData <- data.frame('ID' = rownames(gene.assayData$exprs))
rownames(gene.featureData) = gene.featureData$ID
gene.featureData = gene.featureData %>%
    Biobase::AnnotatedDataFrame(data = ., varMetadata = data.frame('labelDescription' = colnames(.), row.names = colnames(.)))
```

#### Phenotypic data

Biosample metadata

```{r}
metadata = read.table(file.path(Sys.getenv('DBDIR'),
                                'PRJNA239561_metadata_cleaned.tsv'), sep = '\t', header = TRUE)
```

Re-processing stats

```{r}
starLog = read.table(file.path(Sys.getenv('DBDIR'), 'star-rsem',
                               'starLog.final.tsv'), sep = '\t',header=TRUE)
meta1 = data.frame(t(starLog[,-c(1,2)])) %>%
    set_colnames(paste0('starAlignment.',starLog[[1]])) %>%
    set_rownames(colnames(starLog)[-c(1,2)])
meta1$RunId = rownames(meta1)
str(meta1)
```

```{r}
libnames <- colnames(gene.assayData$exprs) %>% as.character()
allMeta <- data.frame('RunId' = libnames) %>%
    plyr::join(y = metadata, by = 'RunId', type = 'left')  %>%
    plyr::join(y = meta1, by = 'RunId', type = 'left') %>%
    set_rownames(libnames)
colDescription = colnames(allMeta)
colDescription[grepl('starAlignment', colnames(allMeta))] = starLog[[2]]
phenoData <-
    Biobase::AnnotatedDataFrame(data = allMeta, varMetadata = data.frame('labelDescription' = colDescription, row.names = colnames(allMeta)))
phenoData@varMetadata
```

Making sure sample order is preserved

```{r}
all(phenoData@data$RunId == colnames(gene.counts))
```

#### Experiment description

Generated earlier, same for all data sets.

#### Assembling an `ExpressionSet`

```{r}
rnor.rnaseq.gene.star_rsem <- ExpressionSet(assayData = gene.assayData,
                             phenoData = phenoData,
                             experimentData = experimentData,
                             featureData = gene.featureData,
                             annotation = 'Ensembl Rnor_6.0.99 genome + ERCC-92'
                             )
```

### Transcript-level `ExpressionSet`

#### Assay data

```{r}
transcript.counts = readRDS(file.path(Sys.getenv('DBDIR'), 'star-rsem', 'matrix.transcripts.expected_count.RDS'))
transcript.tpm = readRDS(file.path(Sys.getenv('DBDIR'), 'star-rsem', 'matrix.transcripts.tpm.RDS'))
transcript.assayData <- new.env()
assign('exprs', transcript.counts, envir = transcript.assayData) # exprs is required
assign('expected_count', transcript.counts, envir = transcript.assayData)
assign('tpm', transcript.tpm, envir = transcript.assayData)
ExpressionSet(assayData = transcript.assayData)
```

#### Phenotypic data

Same as that of gene-level data set.

#### Annotations and features

```{r}
feature_attrs = read.table(file.path(Sys.getenv('DBDIR'), 'star-rsem', 'feature_attrs.rsem.transcripts.tsv'), sep = '\t', header=TRUE, colClasses = 'character')
transcript.featureData <- data.frame('transcript_id' = rownames(transcript.assayData$exprs)) %>%
    plyr::join(y = feature_attrs[, c('transcript_id', 'gene_id', 'length', 'effective_length')], by = 'transcript_id', type = 'left') %>%
    set_names(c('ID', 'GeneId', 'Length', 'EffectiveLength'))
rownames(transcript.featureData) = transcript.featureData$ID
transcript.featureData = transcript.featureData %>%
    Biobase::AnnotatedDataFrame(data = ., varMetadata = data.frame('labelDescription' = colnames(.), row.names = colnames(.)))
```

#### Experiment description

Generated earlier, same for all data sets.

#### Assembling an `ExpressionSet`

```{r}
rnor.rnaseq.transcript.star_rsem <- ExpressionSet(assayData = transcript.assayData,
                             phenoData = phenoData,
                             experimentData = experimentData,
                             featureData = transcript.featureData,
                             annotation = 'Ensembl Rnor_6.0.99 genome + ERCC-92'
                             )
```

## Writing data sets

```{r,eval=FALSE}
usethis::use_data(rnor.rnaseq.gene.kallisto,internal = FALSE, overwrite = TRUE)
usethis::use_data(rnor.rnaseq.transcript.kallisto,internal = FALSE, overwrite = TRUE)
usethis::use_data(rnor.rnaseq.gene.star_rsem, internal = FALSE, overwrite = TRUE)
usethis::use_data(rnor.rnaseq.transcript.star_rsem,internal = FALSE, overwrite = TRUE)
```
